pkg/keys: make PrettyPrintRange redactable

aa-joshi opened the pull request:
This PR is split into the below 3 commits. Hopefully that should make it easier to review.

---

pkg/keys: make PrettyPrintRange redactable

Currently, in Span, RSpan and RangeDescriptor, the start and end keys
are already redactable. But the current implementation of
PrettyPrintRange ignores the redactable nature of the keys. The markers
are stripped and raw strings are used to form the string representation
of the Span/RSpan/RangeDescriptor. This led to the entire string getting
redacted.

This commit fixes that by incorporating the redaction markers into the
algorithm that constructs the string representation in PrettyPrintRange
and returns a `redact.RedactableString`. This will make sure that only
the sensitive parts of the keys are redacted.

Summary of changes:

  * PrettyPrintRange returns RedactableString instead of standard string
  * The CopyEscape function (now renamed to CopyEscapeTrunc) is now
    responsible for maintaing the redactability while copying contents.
  * CopyEscapeTrunc also handles truncating the string if it exceeds the
    maxChars
  * Export CopyEscapeTrunc so that unit tests can be written for it
    (since tests are in keys_test package)
  * Fix tests that expect overly redact keys/ranges

Epic: CRDB-37533
Part of: CRDB-44885
Release note: None

---

pkg/kv/kvserver/concurrency: update test fixtures

The previous commit fixes the redactability of the `PrettyPrintRange`
function. This commit updates the test fixtures used in
`TestConcurrencyManagerBasic` according to the new format of
`PrettyPrintRange`, where only the sensitive part of a range is
redacted.

Epic: CRDB-37533
Part of: CRDB-44885
Release note: None

---

pkg/roachpb: implement `SafeFormatter` in `LockAcquisition`

The previous commit implemented the `SafeFormat` method in `Span`. `Span` is
embedded in the `LockAcquisition` struct. As a result, `LockAcquisition`
inherits the `SafeFormat` method of `Span`, which is not the desired behavior.

This commit implements the `SafeFormat` method in `LockAcquisition` to override
the `SafeFormat` method of `Span.`

Epic: CRDB-37533
Part of: CRDB-44885
Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/138632)
<!-- Reviewable:end -->

aa-joshi reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/138632)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained

___
*[`pkg/keys/printer.go` line 874 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/138632#-OGe2SYu0H5_Sj2q2QX1:-OGe2SYu0H5_Sj2q2QX2:bexyf8d) ([raw file](https://github.com/cockroachdb/cockroach/blob/06f42d623fc739b2eb3c4a01b08eea47fe6ca0eb/pkg/keys/printer.go#L874)):*
> ```Go
> // It prints at most maxChars, truncating components as needed. See
> // TestPrettyPrintRange for some examples.
> func PrettyPrintRange(start, end roachpb.Key, maxChars int, redact bool) string {
> ```

Reason for "why we introduce bool flag for redact instead of returning Redactable string"?
Ans:  Our primary is to combine 2 redactable strings of start & end key into single redactable string. Keys can be dynamic. This means that for a given a start key in the range, the redaction markers could be at different places than the end key. So this would create multiple combination of cases in PrettyRange. To keep it straightforward, we introduce flag for redact. If it is true then we are redacting PII at the source so that prettyprint would be simple.


<!-- Sent from Reviewable.io -->

aa-joshi commented on pkg/keys/printer.go:
```diff
@@ -871,12 +871,19 @@ func init() {
 // It prints at most maxChars, truncating components as needed. See
 // TestPrettyPrintRange for some examples.
-func PrettyPrintRange(start, end roachpb.Key, maxChars int) string {
+func PrettyPrintRange(start, end roachpb.Key, maxChars int, redact bool) string {
```
Initial plan was to return RedactableString. The main challenge StartKey & EndKey are RedactableStrings. So We need to merge 2 RedactableString objects into 1 RedactableString object. The unsafe part of key can be dynamic. So merging 2 RedactableString is tricky. Hence, I am taking redact as bool flag so that we can decide whether we need to redact unsafe part or not. Then we can merge 2 strings. If anyone has better suggestion, I am happy to discuss.

rafiss replied:
> So merging 2 RedactableString is tricky.

were you aware of the `redact.StringBuilder` type? it makes it pretty easy to combine two RedactableStrings -- just pass them in using `(redact.StringBuilder).Printf`

here's an example:
https://github.com/cockroachdb/cockroach/blob/c83c57d354741ac36740894f7387c014eb6c09fe/pkg/kv/kvclient/kvcoord/txn_interceptor_span_refresher.go#L544

the approach of using `Redact()` followed by `StripMarkers()` is not ideal, since it result in output like what you had in that test below, where the redacted content (`×`) is not enclosed in any redaction markers.

```
	Raw:r1179:/Table/53/1/×{-}
```
<!-- Reviewable comment -OHPUBTM51WgPQKz8ylk:buro7j8 -->
<!-- Sent from Reviewable.io -->

rafiss reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/138632)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @aa-joshi and @arjunmahishi)



<!-- Sent from Reviewable.io -->

rafiss reviewed (commented):
since most of the usages here are in KV code, i'll tag @arulajmani as the KV tech lead for awareness, in case you want someone on your team to give this a review.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/138632#-:-OHPVDtY770oYL3oRan6:b-3tjwa3)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @aa-joshi and @arjunmahishi)



<!-- Sent from Reviewable.io -->

arjunmahishi commented:
Hi @rafiss / @arulajmani, I've added a commit that addresses the comments. The `PrettyPrintRange` now returns a `RedactableString` instead of standard strings. Can you please take a look when you get time? 

Thanks!

aa-joshi reviewed (commented):
Thanks for checking this!
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/138632#-:-OIjAjchCJlRMQp870jC:b5axk6o)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @arjunmahishi, @arulajmani, and @dhartunian)



<!-- Sent from Reviewable.io -->

dhartunian reviewed (commented):
Please update the PR message to include the additional details in second commit message.

> Reviewed 3 of 4 files at r1, 3 of 3 files at r4.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/138632#-:-OIlBdwJ8CD2tWCmRpJM:bswq2wi)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @aa-joshi, @arjunmahishi, and @arulajmani)

___
*[`-- commits` line 54 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/138632#-OIlIEQK0us_dGD0jm_L:-OIlIEQK0us_dGD0jm_M:buyd2s0):*
@arjunmahishi @aa-joshi please avoid mixing your work in a single PR because then it creates weird situation with reviews because the author of the diff can now approve.

I reviewed commit 1, and then saw my comments become invalid in the second commit. Please be mindful of the reviewer's time. I always look at things commit by commit because I assume that each contains an atomic unit of work.

I think in this case, the two commits should be squashed, correct?
___
*[`pkg/keys/printer.go` line 874 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/138632#-OIlK08957cLqDVyYc9y:-OIlK08957cLqDVyYc9z:b-unw62f) ([raw file](https://github.com/cockroachdb/cockroach/blob/14307ad2f28db81e58b4b828f81d081effaf6d85/pkg/keys/printer.go#L874)):*
> ```Go
> //
> // It prints at most maxChars, truncating components as needed. See
> // TestPrettyPrintRange for some examples.
> ```

Please update docstring to describe how this function behaves.
___
*[`pkg/keys/printer.go` line 934 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/138632#-OIlKk5jBfb38nwc8aw6:-OIlKk5jBfb38nwc8aw7:bz1jwrc) ([raw file](https://github.com/cockroachdb/cockroach/blob/14307ad2f28db81e58b4b828f81d081effaf6d85/pkg/keys/printer.go#L934)):*
> ```Go
> 	}
> 
> 	buf.Grow(maxChars)
> ```

We often pass maxint here. We do not want to allocate a gigantic buffer.
___
*[`pkg/keys/printer.go` line 946 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/138632#-OIlL_f7888vI2hBZ5Fa:-OIlL_f7888vI2hBZ5Fb:b-apilp8) ([raw file](https://github.com/cockroachdb/cockroach/blob/14307ad2f28db81e58b4b828f81d081effaf6d85/pkg/keys/printer.go#L946)):*
> ```Go
> 			break
> 		}
> 	}
> ```

I don't understand how `safe` is set here. If the redaction marker comes later in the string, shouldn't `safe` be false?
___
*[`pkg/keys/printer.go` line 964 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/138632#-OIlLHpY4ykGlxcBGkA6:-OIlLHpY4ykGlxcBGkA7:b6eqhh) ([raw file](https://github.com/cockroachdb/cockroach/blob/14307ad2f28db81e58b4b828f81d081effaf6d85/pkg/keys/printer.go#L964)):*
> ```Go
> 
> 		buf.UnsafeByte(c)
> 	}
> ```

why do we need to create these functions inline?


<!-- Sent from Reviewable.io -->

dhartunian commented:
Also, lots of tests are failing. Some appear to be redaction related, some not.

arjunmahishi committed:
pkg/keys: make PrettyPrintRange redactable

Currently, in Span, RSpan and RangeDescriptor, the start and end keys
are already redactable. But the current implementation of
PrettyPrintRange ignores the redactable nature of the keys. The markers
are stripped and raw strings are used to form the string representation
of the Span/RSpan/RangeDescriptor. This led to the entire string getting
redacted.

This commit fixes that by incorporating the redaction markers into the
algorithm that constructs the string representation in PrettyPrintRange
and returns a `redact.RedactableString`. This will make sure that only
the sensitive parts of the keys are redacted.

Summary of changes:

  * PrettyPrintRange returns RedactableString instead of standard string
  * The CopyEscape function (now renamed to CopyEscapeTrunc) is now
    responsible for maintaing the redactability while copying contents.
  * CopyEscapeTrunc also handles truncating the string if it exceeds the
    maxChars
  * Export CopyEscapeTrunc so that unit tests can be written for it
    (since tests are in keys_test package)
  * Fix tests that expect overly redact keys/ranges

Epic: CRDB-37533
Part of: CRDB-44885
Release note: None

arjunmahishi committed:
pkg/roachpb: implement SafeFormatter in LockAcquisition

The previous commit implemented the SafeFormat method in Span. Span is
embedded in the LockAcquisition struct. As a result, LockAcquisition
inherits the SafeFormat method of Span, which is not the desired behavior.

This commit implements the SafeFormat method in LockAcquisition to override
the SafeFormat method of Span.

Epic: CRDB-37533
Part of: CRDB-44885
Release note: None

arjunmahishi committed:
pkg/kv/kvserver/concurrency: update test fixtures

The previous commit fixes the redactability of the PrettyPrintRange
function. This commit updates the test fixtures used in
TestConcurrencyManagerBasic according to the new format of
PrettyPrintRange, where only the sensitive part of a range is
redacted.

Epic: CRDB-37533
Part of: CRDB-44885
Release note: None

arjunmahishi commented on pkg/cli/cli_debug_test.go:
```diff
@@ -26,7 +26,7 @@ func Example_debug_decode_key_value() {
 	//
 	// 0.987654321,0 /Local/Range/Table/53/1/-4560243296450227838/RangeDescriptor (0x016b12bd8980c0b6c2e211ba518200017264736300000000003ade68b109): [/Table/53/1/-4560243296450227838, /Table/53/1/-4559358311118345834)
-	// 	Raw:r1179:/Table/53/1/-45{60243296450227838-59358311118345834} [(n1,s1):1, (n4,s4):2, (n2,s2):4, next=5, gen=4]
+	//	Raw:r1179:/Table/53/1/×{-} [(n1,s1):1, (n4,s4):2, (n2,s2):4, next=5, gen=4]
```
1. Isn't this redacting more than before? or did the log config change in the test?
2. The redaction markers `‹ ›` are missing here (as mentioned in the above comment). Not sure if this is a valid format for redaction. Also, what would this look like when the log config is `redact: false, redactable:true`? The expectation is that it should enclose the actual unredacted content with  `‹ ... ›`
<!-- Reviewable comment -OHLh0cy3S7m4NVJ_Rj5:b22a89 -->
<!-- Sent from Reviewable.io -->

arjunmahishi reviewed (commented):
> Your pull request contains more than 1000 changes. It is strongly encouraged to split big PRs into smaller chunks.

I don't think splitting this into multiple PRs will be possible. The actual code changes is pretty small. Most of the diff is just updating the fixtures of tests.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/138632#-:-OIuXr9a9KwOVkYro4F0:b-py19a6)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @aa-joshi, @arulajmani, @dhartunian, and @jbowens)

___
*[`-- commits` line 54 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/138632#-OIlIEQK0us_dGD0jm_L:-OIuRl6fEAsqZ5LGWTFk:b-myxspl):*
Extremely sorry about the confusion. Will keep PR cleaner in the future.

> I think in this case, the two commits should be squashed, correct?

Yes. I've squashed the two commits that were previously there. Fixing the failing tests resulted in more diff. I've split those changes into two more commits to isolate related changes. The commit messages (updated in the PR description) clearly explains what each commit is doing.
___
*[`pkg/keys/printer.go` line 934 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/138632#-OIlKk5jBfb38nwc8aw6:-OIuSXY0EKa0-INS6bVv:bo6x6fy) ([raw file](https://github.com/cockroachdb/cockroach/blob/14307ad2f28db81e58b4b828f81d081effaf6d85/pkg/keys/printer.go#L934)):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

We often pass maxint here. We do not want to allocate a gigantic buffer.
</blockquote></details>

At the start of this function, we are resetting the value of `maxChars` when it's bigger than `len(s)` 
___
*[`pkg/keys/printer.go` line 946 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/138632#-OIlL_f7888vI2hBZ5Fa:-OIuT5cA9ATSkw0tB1R6:b-aigy7w) ([raw file](https://github.com/cockroachdb/cockroach/blob/14307ad2f28db81e58b4b828f81d081effaf6d85/pkg/keys/printer.go#L946)):*
> If the redaction marker comes later in the string, shouldn't `safe` be false

Yes, If we encounter an open `›` later in the string, this sets `safe` to `false` because all chars until `›` are unsafe. The other possible cases are:
- No redaction marker - Everything is safe
- Finding `‹`later in the string  - Everything until `‹` is safe

means the starting start should be `safe=true`

---

I've added a comment with the same explanation.
___
*[`pkg/keys/printer.go` line 964 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/138632#-OIlLHpY4ykGlxcBGkA6:-OIuWQKqCXhZ_Qn7StOZ:b-g2ply9) ([raw file](https://github.com/cockroachdb/cockroach/blob/14307ad2f28db81e58b4b828f81d081effaf6d85/pkg/keys/printer.go#L964)):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

why do we need to create these functions inline?
</blockquote></details>

The `printStr` and `printByte` functions are defined as closures to access the `buf` and `safe` from the outer scope. And these are only called from within `CopyEscapeTrunc`. 

---

Is there any downside to this? There should be no performance over head right?
___
*[`pkg/keys/printer.go` line 882 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/138632#-OHLfZN48RBpZ1OEY6Gj:-OHLfZN48RBpZ1OEY6Gk:buxjlew) ([raw file](https://github.com/cockroachdb/cockroach/blob/96a39144d666bffe06eef4893bb342b7f9b2da42/pkg/keys/printer.go#L882)):*
> ```Go
> 	var prettyStart string
> 	if redact {
> 		prettyStart = redactablePrettyStart.Redact().StripMarkers()
> ```

I'm a little skeptical about this. `.Redact()` will lead to `‹×›`. Then calling `.StripMarkers()` will lead to just `×`. Is this the intent?


<!-- Sent from Reviewable.io -->

arjunmahishi reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/138632)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @aa-joshi, @arulajmani, @dhartunian, @jbowens, and @rafiss)

___
*[`pkg/keys/printer.go` line 874 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/138632#-OIlK08957cLqDVyYc9y:-OIuaNaL086ndsipztGg:bb6un1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/14307ad2f28db81e58b4b828f81d081effaf6d85/pkg/keys/printer.go#L874)):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

Please update docstring to describe how this function behaves.
</blockquote></details>

Done


<!-- Sent from Reviewable.io -->

arjunmahishi commented:
Found a way to split this into separate PRs. Closing this PR as it's very unreviewable
